core/src/main/scala/kafka/controller/ReassignmentsManager.scala

```diff
@@ -0,0 +1,520 @@
+package kafka.controller
+
+import kafka.admin.AdminOperationException
+import kafka.common.StateChangeFailedException
+import kafka.utils.Logging
+import kafka.zk.{KafkaZkClient, ReassignPartitionsZNode, TopicPartitionStateZNode}
+import kafka.zk.KafkaZkClient.UpdateLeaderAndIsrResult
+import org.apache.kafka.common.errors.ControllerMovedException
+import org.apache.kafka.common.protocol.Errors
+import org.apache.kafka.common.requests.ApiError
+import org.apache.kafka.common.{KafkaException, TopicPartition}
+import org.apache.zookeeper.KeeperException
+import org.apache.zookeeper.KeeperException.Code
+
+import scala.collection.{Map, Seq, Set, immutable, mutable}
+
+/**
+ * A helper class which contains logic for driving partition reassignments.
+ * This class is not thread-safe.
+ */
+class ReassignmentsManager(controllerContext: ControllerContext,
+                           zkClient: KafkaZkClient,
+                           topicDeletionManager: TopicDeletionManager,
+                           replicaStateMachine: ReplicaStateMachine,
+                           partitionStateMachine: PartitionStateMachine,
+                           eventManager: ControllerEventManager,
```

Comment:
I don't really like passing in the eventManager here. It is used in 
1. when we remove the `/reassign_partitions` znode
2. when we register a new ISR ZNodeChangeHandler

We could circumvent this by passing in two methods `onZNodeDeletion` and `onReassignmentStart` depending on people's thoughts

Reply:
I like the your idea of adding callbacks. I think we can probably turn it into a Listener or something. For example:
```scala
trait ReassignmentListener {
  def onReassignmentUpdated // invoked in `updateCurrentReassignment`
  def onReassignmentResumed // invoked at the start of `onPartitionReassignment`
  def onReassignmentFinished // invoked at the end of `onPartitionReassignment` (case B)
}
```
Using this approach, we can probably also get rid of the dependence on `TopicDeletionManager`.